incubator/orm/auto_uint64.go

```diff
@@ -1,49 +1,216 @@
-	return autoUInt64TableBuilder{prefix: prefix, key: key, cdc: cdc}
+const ormCodespace = "orm"
+
+var (
+	// todo: ormCodespace ok or do we need to claim error codes somehow?
+	ErrNotFound         = errors.Register(ormCodespace, 100, "not found")
+	ErrIteratorDone     = errors.Register(ormCodespace, 101, "iterator done")
+	ErrType             = errors.Register(ormCodespace, 102, "invalid type")
+	ErrUniqueConstraint = errors.Register(ormCodespace, 103, "unique constraint violation")
+)
+
+var _ TableBuilder = &autoUInt64TableBuilder{}
+
+func NewAutoUInt64TableBuilder(prefix []byte, key sdk.StoreKey, cdc *codec.Codec, model interface{}) *autoUInt64TableBuilder {
+	if len(prefix) == 0 {
+		panic("prefix must not be empty")
+	}
+	if cdc == nil {
+		panic("codec must not be empty")
+	}
+	if model == nil {
+		panic("model must not be empty")
+	}
+	tp := reflect.TypeOf(model)
+	if tp.Kind() == reflect.Ptr {
+		tp = tp.Elem()
+	}
+	return &autoUInt64TableBuilder{prefix: prefix, storeKey: key, cdc: cdc, model: tp}
 }
 
 type autoUInt64TableBuilder struct {
-	prefix      byte
-	key         sdk.StoreKey
+	model       reflect.Type
+	prefix      []byte
+	storeKey    sdk.StoreKey
 	cdc         *codec.Codec
-	indexerRefs []indexRef
+	afterSave   []AfterSaveInterceptor
+	afterDelete []AfterDeleteInterceptor
+}
+
+// todo: this function gives access to the storage. It does not really fit the builder patter.
+func (a autoUInt64TableBuilder) RowGetter() RowGetter {
+	return func(ctx HasKVStore, rowId uint64, dest interface{}) ([]byte, error) {
+		store := prefix.NewStore(ctx.KVStore(a.storeKey), a.prefix)
+		key := EncodeSequence(rowId)
+		val := store.Get(key)
+		// todo: how to handle not found?
+		if val == nil {
+			return nil, ErrNotFound // todo: discuss how to handle this scenario if we drop error return parameter
+		}
+		return key, a.cdc.UnmarshalBinaryBare(val, dest)
+	}
+}
+
+func (a autoUInt64TableBuilder) StoreKey() sdk.StoreKey {
+	return a.storeKey
 }
 
-func (a autoUInt64TableBuilder) RegisterIndexer(prefix byte, indexer Indexer) {
-	a.indexerRefs = append(a.indexerRefs, indexRef{prefix: prefix, indexer: indexer})
+func (a autoUInt64TableBuilder) Build() autoUInt64Table {
+	seq := NewSequence(a.storeKey, a.prefix)
+	return autoUInt64Table{
+		model:       a.model,
+		sequence:    seq,
+		prefix:      a.prefix,
+		storeKey:    a.storeKey,
+		cdc:         a.cdc,
+		afterSave:   a.afterSave,
+		afterDelete: a.afterDelete,
+	}
 }
 
-func (a autoUInt64TableBuilder) Build() AutoUInt64Table {
-	return autoUInt64Table{}
+func (a *autoUInt64TableBuilder) AddAfterSaveInterceptor(interceptor AfterSaveInterceptor) {
+	a.afterSave = append(a.afterSave, interceptor)
 }
 
+func (a *autoUInt64TableBuilder) AddAfterDeleteInterceptor(interceptor AfterDeleteInterceptor) {
+	a.afterDelete = append(a.afterDelete, interceptor)
+}
+
+var _ AutoUInt64Table = autoUInt64Table{}
```

Comment:
Same point about making this public. Why forcing this private struct to an interface, rather than just making it public (with all private members)